kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/types/LeastCommonTypeResolver.java

```diff
@@ -0,0 +1,284 @@
+                return BinaryType.BINARY;
+            } else if (hasGroup5) {
+                return BooleanType.BOOLEAN;
+            } else if (hasGroup6) {
+                throw new IllegalStateException(
+                        "Interval is not supported yet for least common type resolution");
+            } else if (hasGroup7) {
+                throw new IllegalStateException(
+                        "ArrayType is not supported yet for least common type resolution");
+            } else if (hasGroup8) {
+                throw new IllegalStateException(
+                        "MapType is not supported yet for least common type resolution");
+            } else if (hasGroup9) {
+                throw new IllegalStateException(
+                        "StructType is not supported yet for least common type resolution");
+            } else {
+                throw new IllegalStateException(
+                        "There are no least common type resolution");
+            }
+        } else {
+            throw new IllegalStateException("There are no least common type resolution");
+        }
+    }
+
+    protected static DataType getDateTypeByPrecedence(byte precedence) {
+        switch (precedence) {
+            case PromotionGroup.TIME_PRECEDENCE_DATE:
+                return DateType.DATE;
+            case PromotionGroup.TIME_PRECEDENCE_TIMESTAMP:
+                return TimestampType.TIMESTAMP;
+            default:
+                String msg = String.format(
+                        "`%s` is not supported for least common type " +
+                                "resolution for time related type`",
+                        precedence);
+                throw new IllegalStateException(msg);
+        }
+    }
+
+    protected static DataType getNumberTypeByPrecedence(byte precedence) {
+        switch (precedence) {
+            case PromotionGroup.NUMBER_PRECEDENCE_BYTE:
+                return ByteType.BYTE;
+            case PromotionGroup.NUMBER_PRECEDENCE_SHORT:
+                return ShortType.SHORT;
+            case PromotionGroup.NUMBER_PRECEDENCE_INT:
+                return IntegerType.INTEGER;
+            case PromotionGroup.NUMBER_PRECEDENCE_LONG:
+                return LongType.LONG;
+            case PromotionGroup.NUMBER_PRECEDENCE_DECIMAL:
+                return DecimalType.USER_DEFAULT;
+            case PromotionGroup.NUMBER_PRECEDENCE_FLOAT:
+                return FloatType.FLOAT;
+            case PromotionGroup.NUMBER_PRECEDENCE_DOUBLE:
+                return DoubleType.DOUBLE;
+            default:
+                String msg = String.format(
+                        "`%s` is not supported for least common type " +
+                                "resolution for number related type`",
+                        precedence);
+                throw new IllegalStateException(msg);
+        }
+    }
+    public static boolean isCompatible(DataType a, DataType b) {
+        if (a instanceof ByteType) {
```

Comment:
Having a static immutable set also make sense, it seems like we can't use `set.of` operation due to java version, so I opt to enumerate it here instead of creating a set by adding possible value one by one.